Step3 - 구간 추가 기능 리뷰요청 드립니다!

aksgudwns opened the pull request:
안녕하세요!
매번 리뷰 빠르게 해주시는데, 제가 항상 늦게 올려서 죄송한 마음입니다
더 고민하고, 더 잘해보고 싶은 마음에 계속 수정하다가 조금씩 늦어지는 것 같습니다..
이번 리뷰도 잘 부탁드리겠습니다!

리뷰요청드리면서 궁금한 점이 있어서 두가지 문의드립니다!
1. 작업하다보니, LineService에서 Line생성시 SectionService를 통해 Section을 생성하고, 
SectionService에서 Section생성시 LineService에서 lineId로 Line정보를 조회하면서 순환참조가 일어났습니다.
우선 저는 SectionLineStationService와 SectionStationService를 만들어 순환참조관계를 풀어보았습니다.
제가 설계를 잘못한건지, 아니면 더 좋은 방법이 있는지 같이 문의 드리겠습니다

2. RestAssured를 통해 오류케이스 테스트를 진행하려 했지만, 
api response에는 internal server error만 담기고, jdk에서 발생한 오류는 알 수 없기에 아래 클래스를 통하여 Service테스트를 진행하였습니다. (src/test/java/nextstep/subway/section/SectionLineStationServiceTest.java)
이 방법으로 하는게 올바른지, 더 좋은 방법이 있는지 궁금합니다.
api 응답에 500에러 체크만으로도 충분할까요?

감사합니다:)

aksgudwns committed:
doc - 마크다운 수정

aksgudwns committed:
refactor - 리뷰반영

- getStationById StationService로 이동
- domain에서 dto에 의존하지 않도록 수정
- containsExactly활용으로 테스트 메서드 가독성 향상
- 테스트메서드에서 //then, //when 확실히 구분하여 가독성 향상

aksgudwns committed:
feat - 역 사이에 새로운 구간 등록 Happy Path

aksgudwns committed:
fix - Line등록,삭제시 LineStation 테이블 등록,삭제 문제 수정, containsEactly테스트메소드 수정

- Line 저장시 LineStation 테이블에 저장
- Line 삭제시 LineStation 테이블에서 데이터 삭제
- containsEactly set으로 변환시 순서가 틀어져서 containsOnly로 수정

aksgudwns committed:
refactor - LineStation -> Section으로 대체 및 순환참조 방지를 위한 신규 서비스 생성

- LineStation -> Section으로 대체
- 순환참조 방지를 위한 SectionStationService, LineStationService 추가
- atdd 테스트 추가 상위역으로 새로운역을 넣는 케이스와 하위역으로 새로운 역을 넣는 케이스 작성

aksgudwns committed:
feat - 상행역과 하행역이 모두 등록되어 있는 경우 테스트 추가

aksgudwns committed:
feat - 상행역,하행역 전부 노선에 존재하지 않는 경우 기능구현 테스트 추가, 테스트메서드 리팩토링

- 상행역,하행역 전부 존재하지 않는경우 테스트 및 기능구현
- stationId를 long 타입으로 리턴하도록 수정

aksgudwns committed:
feat - 섹션 구간 길이 오류 테스트 및 기능 구현

- 역 사이에 새로운 역을 등록할 경우 기존 역 사이 길이보다 크거나 같은 경우 등록 x
- 길이 값이 0보다 작거나 같은 경우 등록 x

aksgudwns committed:
feat - 상행종점, 하행종점 등록테스트 추가

seokhongkim commented on README.md:
```diff
@@ -62,22 +62,29 @@ This project is [MIT](https://github.com/next-step/atdd-subway-admin/blob/master
- - When 생성한 지하철 노선을 삭제하면
- - Then 해당 지하철 노선 정보는 삭제된다.
\ No newline at end of file
+- [x] 지하철노선 생성 
```
기능 요구 사항 잘 정리해주셨습니다. 👍

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -33,31 +35,29 @@ public List<LineResponse> findAllLines() {
                 .orElseThrow(() -> new NoSuchElementException(ErrorMessage.LINE_NO_FIND_BY_ID.getMessage()));
     }
 
+    public LineResponse findLineById(Long id) {
```
`getLineById`와 `findLineById`는 get, find의 차이인데 반환 값의 타입이 달라지네요. find, get이 아닌 다른 표현으로 명시적으로 구분을 해보시면 좋겠습니다.
find, get은 메서드의 특정 행위를 구분하기 위해 사용하는 경우가 있습니다. https://tuhrig.de/find-vs-get 블로그에 있는 내용을 참고해보셔도 좋겠습니다.

aksgudwns replied:
참고자료 공유해주셔서 감사합니다!
앞으로는 get과 find 구분해서 사용하겠습니다!
내용이 없는 경우 exception이 발생하므로 get으로 통일하였습니다

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -33,31 +35,29 @@ public List<LineResponse> findAllLines() {
         Line line = lineRepository.findById(id)
                 .orElseThrow(() -> new NoSuchElementException(ErrorMessage.LINE_NO_FIND_BY_ID.getMessage()));
-        line.updateRequest(lineUpdateRequest);
+        line.updateNameAndColor(lineUpdateRequest.getName(), lineUpdateRequest.getColor());
```
메서드 이름을 구체적으로 변경하고 dto 에 의존하지 않도록 수정하신 것 좋습니다. 👍

seokhongkim commented on src/main/java/nextstep/subway/application/SectionLineStationService.java:
```diff
@@ -0,0 +1,94 @@
+        this.stationService = stationService;
+    }
+
+    public SectionsResponse addSection(Long lineId, SectionRequest sectionRequest) {
```
`upStation`, `downStation` 을 이용해서 `Section` 을 만들고, `Line` 에게 `Section` 을 추가하라고 요청해보는 건 어떨까요?
`Line` 내부에서 `Sections` 에게 이 기능을 위임하면, 서비스의 로직을 줄이면서 `Sections` 가 도메인 로직을 관리하도록 수정할 수 있을 것 같습니다.
그럼 `SectionLineStationService` 도 제거할 수 있겠네요. :)

aksgudwns replied:
말씀들어보니, 서비스에 너무 많은 로직이 있었네요..
도메인에서 로직관리하도록 변경해보겠습니다!

seokhongkim commented on src/main/java/nextstep/subway/application/SectionStationService.java:
```diff
@@ -0,0 +1,37 @@
+        this.stationService = stationService;
+    }
+
+    public void initLineSectionAndAddSections(Line persistLine, LineRequest lineRequest) {
```
이 로직도 `Line`을 만들 때 필요한 `Section`을 만든 후, 생성하는 시점에 `Sections`를 통해 처리되도록 구현해보셔도 좋겠습니다.

aksgudwns replied:
이곳 로직도 도메인에서 관리하도록 수정하겠습니다!

seokhongkim commented on src/main/java/nextstep/subway/application/SectionStationService.java:
```diff
@@ -0,0 +1,37 @@
+                new Section(persistLine, upStation, downStation, lineRequest.getDistance())));
+    }
+
+    public void deleteSectionsByLine(Line line) {
```
`Line`에게 삭제하라고 요청을 해보면 어떨까요? JPA 미션 때 해보신 것처럼, `Line`이 `Sections`에게 삭제를 요청하고, `Line`도 삭제 상태로 변할 수 있지는 않을까요?
또는 연관관계 설정에 있는 `CascadeType`을 이용해서 삭제를 전파할 수도 있겠네요.

aksgudwns replied:
CascadeType추가하여 필요없는 repository, service삭제하겠습니다!

seokhongkim commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -19,25 +15,18 @@ public class Line extends BaseEntity {
     }
 
-    public Line(String name, String color, Station upStation, Station downStation, int distance) {
+    public Line(String name, String color, int distance) {
```
`distance`도 삭제해도 될 것 같습니다.

aksgudwns replied:
distance는 삭제하고, 라인 길이검증은 sections내부 메서드를 이용하겠습니다.

seokhongkim commented on src/main/java/nextstep/subway/domain/SectionRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface SectionRepository extends JpaRepository<Section, Long> {
```
`Line`을 조회해오면 `Sections`를 모두 가지고 올 수 있으니, `Sections`를 통해 `Section`을 검색하고 삭제하는 기능을 구현해보는 건 어떨까요?

aksgudwns replied:
도메인으로 로직을 옮기고, CascadeType을 사용하니, SectionRepository가 필요없어졌습니다.
삭제하겠습니다!

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,65 @@
+        return stations;
+    }
+
+    public void add(Section section) {
```
구간을 추가할 때 필요한 검증을 `Sections`, `Section`을 이용해서 여기서 구현해보시면 좋겠네요!

aksgudwns replied:
검증하는 부분도 도메인단 로직으로 수정하겠습니다:)

seokhongkim commented on src/test/java/nextstep/subway/section/SectionLineStationServiceTest.java:
```diff
@@ -0,0 +1,120 @@
+    @DisplayName("상행역과 하행역이 모두 구간에 이미 등록되어 있는 경우 오류 발생 테스트")
+    void upStationDownStationAlreadyExistExceptionTest() {
+        assertThatThrownBy(() -> sectionLineStationService
+                .addSection(lineId, new SectionRequest(upStationId,downStationId,10)))
```
도메인 로직을 `Sections`, `Section`에 구현하면 DB에 대한 의존 없이도 데이터를 주입해서 단위 테스트를 해보실 수 있을 것 같습니다. :)
프로덕션 코드 리팩터링 이후 테스트도 리팩터링 해보시면 좋겠네요!

aksgudwns replied:
리팩토링하니, DB의존없이 원하는 경우를 전부 테스트할 수 있게 되었네요:)
이름도 SectionsTest로 수정하였습니다!

seokhongkim reviewed (changes requested):
형준님, 3단계 미션 잘 진행해주셨습니다. 👍 

먼저 질문주신 부분에 대해 제 생각 말씀드려요.
> 작업하다보니, LineService에서 Line생성시 SectionService를 통해 Section을 생성하고,
SectionService에서 Section생성시 LineService에서 lineId로 Line정보를 조회하면서 순환참조가 일어났습니다.
우선 저는 SectionLineStationService와 SectionStationService를 만들어 순환참조관계를 풀어보았습니다.
제가 설계를 잘못한건지, 아니면 더 좋은 방법이 있는지 같이 문의 드리겠습니다.

순환 참조를 발견하고 풀기 위해 고민해보고 시도해보신 점 좋습니다.
서로 참조하고 있는 엔티티에 대해 Service 에서 많은 로직을 가지고 있어서 풀어가는 것이 조금 더 어려우셨을 수도 있을 것 같네요.
서비스와 엔티티에 남긴 코멘트를 참고하셔서 비즈니스 로직을 이동하면 자연스럽게 풀어보실 수 있을 것 같습니다.

--- 
> RestAssured를 통해 오류케이스 테스트를 진행하려 했지만,
api response에는 internal server error만 담기고, jdk에서 발생한 오류는 알 수 없기에 아래 클래스를 통하여 Service테스트를 진행하였습니다. (src/test/java/nextstep/subway/section/SectionLineStationServiceTest.java)
이 방법으로 하는게 올바른지, 더 좋은 방법이 있는지 궁금합니다.
api 응답에 500에러 체크만으로도 충분할까요?

각 테스트의 작성 목적과 검증해야 할 범위를 먼저 고민해보시면 도움이 될 것 같습니다.
저는 말씀하신 코드 중간에 발생하는 예외의 타입이 인수 테스트의 관심사는 아니라고 생각합니다. 따라서 500 또는 예외 상황에 맞는 적절한 코드가 발생하는지, 또는 적절한 예외 메시지가 응답으로 오는지를 검증할 것 같습니다.
단위 테스트나 통합 테스트(서비스 + DB)를 테스트 범위로 진행하는 경우에는 발생하는 예외의 타입과 메세지에도 관심을 가질 것 같습니다.

---

이외에 코드를 보며 궁금한 점과 고민해보시면 도움이 될 것 같은 내용을 코멘트로 남겼습니다.
코멘트 확인 부탁드립니다. 🙏

aksgudwns committed:
refactor - 서비스로직 도메인으로 변경, cascadeType추가

- 서비스로직 도메인으로 변경
- CascadeType 추가
- SectionStationService 등 불필요 서비스, 레퍼지토리 삭제
- find -> get으로 수정

aksgudwns committed:
refactor - distance 필드 삭제, 라인길이 비교 기능 구현

aksgudwns committed:
refactor - SectionLineStationServiceTest 테스트 로직 Sections테스트로 수정

- 도메인으로 비즈니스로직을 변경하면서 테스트클래스 수정

aksgudwns committed:
refactor - Section의 primitivetype distance 클래스로 포장 & orphanRemoval 추가

- Section에 distance int type 래핑클래스 추가
- 자식객체 함께 삭제를 위해 orphanRemoval추가
- distances 추가

aksgudwns committed:
refactor - Distance, Distances 잘못된 위치 수정

aksgudwns committed:
refactor - sections에서 Stations객체 찾을 때 스트림사용 의미 명확하게 수정

aksgudwns commented:
말씀주신 커멘트 반영해보았습니다 다시 리뷰부탁드립니다!
Service에 있는 로직을 도메인으로 옮기고 나니 코드가 한결 깔끔해진 느낌입니다.
Section에 primitive type인 distance는 클래스로 래핑하여 리팩토링 해보았습니다!

한가지 작업하다가 궁금한점이 있어서 문의 드립니다..
Line.java 56~57줄에 Section을 생성하고 후에 setLine으로 Section에 line을 주입하게 작업하였습니다.
setter사용은 지양하는 것으로 알고 있지만.. 생성자만으로 처리할 경우 addAndGetSections메서드에 line을 변수로 추가해야 합니다. 
그러면 파라미터가 너무많아지는 이슈가 생기고, Sections도메인에서 Line에 대한 의존성이 생겨 좋지않은 방법이라 생각하였습니다.
그래서 setter를 사용하여 작업했는데, 혹시 다른 좋은방법이 있을지 문의드립니다!

감사합니다.

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -33,15 +34,19 @@ public List<LineResponse> findAllLines() {
 
-    public LineResponse findLineById(Long id) {
-        return lineRepository.findById(id).map(LineResponse::of)
+    public Line getLineDomainById(Long id) {
```
`Station`을 조회할 때는 `getStationById`고, `Line`의 경우 domain과 dto를 구분하기 위해 `Domain` suffix를 추가하셨군요. :)

이외에도 다른 가능한 방법들을 적어봅니다. 보시고 괜찮은 것 같은 방법이 있다면 참고해서 리팩터링 해보셔도 좋겠습니다.
1. dto로 변환하는 과정을 컨트롤러에서 수행한다. 서비스에서는 도메인 계층에만 의존하도록 하여, 추후 다른 dto가 필요한 경우에도 서비스의 수정 없이 컨트롤러에서 구분한다.
2. dto로 조회하는 것을 `getLineResponseById`로 변경한다. 서비스에서는 `findById`, `getById`를 사용해도 자연스럽게 해당 도메인의 객체를 조회하는 것임을 나타낼 수 있으므로(`Repository` 구현 시 `findBy`, `findAllBy`와 같이 조회 대상 도메인을 생략하는 것과 같음), dto로 조회하는 경우에만 명시적으로 이름을 변경한다.
3. dto로 변경하는 계층을 추가한다. 이는 뷰 레이어와 밀접하므로 뷰를 위한 패키지에서 함께 관리한다.

aksgudwns replied:
넵 말씀주신내용 다음과제 때 반영하겠습니다!

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -50,14 +55,18 @@ public void deleteLineById(Long id) {
     }
 
     private Line lineRequestToLine(LineRequest lineRequest) {
```
이 로직도 `LineRequest`에서 관리해도 괜찮을 것 같네요. :)
`LineResponse`가 `Line`으로부터 `LineResponse`를 생성하는 책임을 가지고 있는 것처럼, `LineRequest`도 `LineRequest`로부터 `Line`을 생성할 수 있을 것 같다는 점에서 의견 드려봅니다.

aksgudwns replied:
말씀들어보니, Dto를 도메인으로 변환해주는 메서드가 Service에 있는게 어색해보이네요..
수정하겠습니다!

seokhongkim commented on src/main/java/nextstep/subway/domain/Distance.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    public Distance(int distance, boolean isZeroAllowed) {
+        if(isZeroAllowed && distance < 0) {
```
`if` 키워드와 소괄호 사이에 공백이 없는 경우가 있네요.

인덴트 설정이 어떻게 되어 있는지 확인해보시면 좋겠습니다.
추가로, 파일을 저장할 때 자동으로 인덴트 조정이 되도록 설정하거나, 파일 수정 후 인덴트 조정을 습관으로 만들어 보시는 것도 좋겠습니다. :)

aksgudwns replied:
앞으로 신경쓰면서 놓치지 않도록 하겠습니다!

seokhongkim commented on src/main/java/nextstep/subway/domain/Distance.java:
```diff
@@ -0,0 +1,56 @@
+        return distance;
+    }
+
+    public Distance getNewSectionDistance(Distance distance) {
```
`dsitance`는 `section`에서 관리되는 객체이므로, `section`에 대한 정보를 모르는 것이 좋다고 생각합니다.
`distance`가 `section`을 아는 순간 순환참조의 형태가 될 수 있습니다.

`distance`는 `minus` 또는 `subtract`와 같은 기능을 제공해주고, 값을 뺀 이후에 계산된 거리가 음수인 경우 예외만 발생시켜보면 어떨까요?

aksgudwns replied:
Section에 대한 정보는 사용하지 않지만.. 다른 사람이 볼때는 네이밍 때문에 오해할 수 있겠군요..
의견주신 부분 맞춰서 수정하겠습니다!

seokhongkim commented on src/main/java/nextstep/subway/domain/Distances.java:
```diff
@@ -0,0 +1,22 @@
+        return Integer.compare(distances.stream().mapToInt(Distance::getDistance).sum(), distance.getDistance());
+    }
+
+    public void checkLessThanAllSectionDistance(Distance distance) {
```
이 밸리데이션은 제거해야 할 것 같습니다.
구간 중간에 역을 추가하는 경우, 거리를 계산하는 과정에서 밸리데이션에 걸려 생성되지 않을 것으로 보입니다. 이게 미션의 요구사항인 것으로 보이네요.
종점으로 추가하는 역의 경우에만 이 밸리데이션으로 영향을 받을텐데, 종점으로 추가하는 역은 기존 구간에 영향을 주지 않으므로 거리를 비교하지 않으셔도 괜찮습니다. :)

aksgudwns replied:
밑에 섹션 생성하는 로직에서 걸러지긴 하지만, 처음에는 터무니 없는 큰 숫자가 들어와서 라인의 전체길이보다 클 경우 처음부터 아예 거르기 위해 로직을 추가하였습니다.
다만 말씀주신 내용 듣고 생각해보니, 자주 발생하는 예외케이스가 아닌 체크로직을 정상적인 케이스에서 매번 수행해야 한다는게 더 불이익이 크다는 생각이 들었습니다.
삭제하도록 하겠습니다!

seokhongkim replied:
중간에 추가되는 역의 경우 라인의 전체 길이보다 큰 것을 빠르게 밸리데이션 해주는 것의 장점이 있다고 생각됩니다. :)
다만, 종점을 추가하는 경우에는 매우 긴 구간이 추가되어도 정상적으로 추가되어야 하기에, 밸리데이션 로직이 버그가 될 수도 있다고 생각합니다.
실제 존재하는 노선을 기준으로 상상해보겠습니다. (현실성은 없지만) 5호선 종점 방화역에서 인천공항역까지 직행 구간이 생기고 이 구간이 기존의 전체 노선 길이보다 길다고 가정 했을 때, 방화 - 인천공항 구간도 추가가 가능해야 하겠네요. 따라서, 기존의 노선 전체 구간보다 긴 구간이 추가되더라도 정상적으로 추가되어야 하는 케이스가 존재할 것 같습니다.

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,129 @@
+        return sections;
+    }
+
+    private Station checkAndGetNewStation(Station upStation, Station downStation) {
```
메서드가 두 가지 역할을 하고 있네요.
1. 밸리데이션
2. 새로 추가하는 역이 상행인지, 하행인지 판단

상행역과 하행역 판단은 별도의 메서드를 통해 판단해보셔도 좋겠습니다.
68번째 라인과 72번째 라인의 `requestUpStation.equals(newStation)`, `requestDownStation.equals(newStation)`을 해당 메서드로 분리하면 이름을 보고 코드를 이해하기 한결 수월할 것 같아요.

aksgudwns replied:
코드가독성 향상을 위해 말씀주신 의견대로 수정해보겠습니다!

seokhongkim commented on src/test/java/nextstep/subway/distance/DistanceTest.java:
```diff
@@ -0,0 +1,58 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class DistanceTest {
```
꼼꼼한 단위 테스트 작성 좋습니다. 👍

aksgudwns replied:
감사합니다 :)

seokhongkim reviewed (approved):
형준님, 3단계 미션 코멘트 잘 반영해주셨습니다. 👍 
3단계 PR도 머지하겠습니다.

수정해주신 코드를 보고 코멘트를 조금 더 남겼습니다. 다음 단계 미션 진행하실 때 참고해보셔도 좋겠습니다.
4단계 미션에서 뵙겠습니다. 😀
